### PR TITLE
Add needed `merge-tree` re-exports to `shared-sequence`

### PR DESCRIPTION
**Pull request by Josmithr**

### CODE REVIEW

**Comment by Josmithr** [comment]
> why is this necessary?

Consumers using the sequence delta event (for example) potentially need to consume the MergeTreeDeltaType. In particular, if they are importing shared-string from the fluid-framework package, they currently have no way to reference that type.

@Abe27342 recommended a few other types here that consumers would need.

File: packages/dds/sequence/src/index.ts [hunk 1]
```diff
@@ -42,3 +42,13 @@ export * from "./sharedObjectSequence";
 export * from "./sharedNumberSequence";
 export * from "./sparsematrix";
 export * from "./sharedIntervalCollection";
+
+export {
```
Comment (Abe27342) [inline packages/dds/sequence/src/index.ts:46]: it's worth splitting out the `export type`s here from the actual exports.

**Comment by anthony-murphy** [comment]
> > why is this necessary?
> 
> Consumers using the sequence delta event (for example) potentially need to consume the MergeTreeDeltaType. In particular, if they are importing SharedString from the fluid-framework package, they currently have no way to reference that type.
> 
> @Abe27342 recommended a few other types here that consumers would need.

why can't they directly use the merge-tree package. in general i don't think this approach will scale. there are lots and lots of concepts that leak from merge-tree to shared string

**Comment by anthony-murphy** [comment]
> > > > why is this necessary?
> > > 
> > > 
> > > Consumers using the sequence delta event (for example) potentially need to consume the MergeTreeDeltaType. In particular, if they are importing SharedString from the fluid-framework package, they currently have no way to reference that type.
> > > @Abe27342 recommended a few other types here that consumers would need.
> > 
> > 
> > why can't they directly use the merge-tree package. in general i don't think this approach will scale. there are lots and lots of concepts that leak from merge-tree to shared string
> 
> I don't think a standard consumer will even know to look for the merge-tree package. Especially if they are using the higher-level fluid-framework package, the expectation is that they will have all the pieces they need to use the library.
> 
> I believe the intention is for the general consumer to use the `fluid-framework` package as a 1-stop shop for everything aside from the backend-specific client implementation(s) they're using. If that's the case, I think we need to ensure consumers have everything they need there.
> 
> As for scalability, API-Extractor has features to help manage this to an extent (by enforcing you explicitly export everything that is transitively referenced by your root API surface). I didn't enable that here because the set of things required from merge-tree to satisfy this is enormous. I suspect a deeper audit would be desirable before tackling that.
> 
> For additional context, I was prototyping something with SharedString last night in a side project I'm working on, and am specifically using the fluid-framework package. In order to actually get my sequence delta event handler working, I had to dig into node_modules to figure out where I needed to get the type from. I don't think that's a reasonable user experience for our customers.

that makes sense, but it seems to me the re-export of shared string is just broken, and i'd suspect other re-export are similarly broken, and we'll continue to introduce new breaks, as the current strategy is not sustainable. the api-extractor  thing seems like a good solution, and the surface area will be enormous.  So i think we should either enable the api-extractor, or remove sequence from fluid-framework, as right now it doesn't really work

**Comment by Josmithr** [comment]
> > > > > why is this necessary?
> > > > 
> > > > 
> > > > Consumers using the sequence delta event (for example) potentially need to consume the MergeTreeDeltaType. In particular, if they are importing SharedString from the fluid-framework package, they currently have no way to reference that type.
> > > > @Abe27342 recommended a few other types here that consumers would need.
> > > 
> > > 
> > > why can't they directly use the merge-tree package. in general i don't think this approach will scale. there are lots and lots of concepts that leak from merge-tree to shared string
> > 
> > 
> > I don't think a standard consumer will even know to look for the merge-tree package. Especially if they are using the higher-level fluid-framework package, the expectation is that they will have all the pieces they need to use the library.
> > I believe the intention is for the general consumer to use the `fluid-framework` package as a 1-stop shop for everything aside from the backend-specific client implementation(s) they're using. If that's the case, I think we need to ensure consumers have everything they need there.
> > As for scalability, API-Extractor has features to help manage this to an extent (by enforcing you explicitly export everything that is transitively referenced by your root API surface). I didn't enable that here because the set of things required from merge-tree to satisfy this is enormous. I suspect a deeper audit would be desirable before tackling that.
> > For additional context, I was prototyping something with SharedString last night in a side project I'm working on, and am specifically using the fluid-framework package. In order to actually get my sequence delta event handler working, I had to dig into node_modules to figure out where I needed to get the type from. I don't think that's a reasonable user experience for our customers.
> 
> that makes sense, but it seems to me the re-export of shared string is just broken, and i'd suspect other re-export are similarly broken, and we'll continue to introduce new breaks, as the current strategy is not sustainable. the api-extractor thing seems like a good solution, and the surface area will be enormous. So i think we should either enable the api-extractor, or remove sequence from fluid-framework, as right now it doesn't really work

I'm actually wondering if Abram's suggestion to explicitly use type exports might reduce the API explosion I initially encountered when I tried the API extractor approach. Let me give that another shot and see if it is reasonable.

And I'm definitely happy to close this out if we want to take a different approach here - more than anything I just wanted to make sure this issue was on our radar 🙂

**Comment by anthony-murphy** [comment]
> > > > > > why is this necessary?
> > > > > 
> > > > > 
> > > > > Consumers using the sequence delta event (for example) potentially need to consume the MergeTreeDeltaType. In particular, if they are importing SharedString from the fluid-framework package, they currently have no way to reference that type.
> > > > > @Abe27342 recommended a few other types here that consumers would need.
> > > > 
> > > > 
> > > > why can't they directly use the merge-tree package. in general i don't think this approach will scale. there are lots and lots of concepts that leak from merge-tree to shared string
> > > 
> > > 
> > > I don't think a standard consumer will even know to look for the merge-tree package. Especially if they are using the higher-level fluid-framework package, the expectation is that they will have all the pieces they need to use the library.
> > > I believe the intention is for the general consumer to use the `fluid-framework` package as a 1-stop shop for everything aside from the backend-specific client implementation(s) they're using. If that's the case, I think we need to ensure consumers have everything they need there.
> > > As for scalability, API-Extractor has features to help manage this to an extent (by enforcing you explicitly export everything that is transitively referenced by your root API surface). I didn't enable that here because the set of things required from merge-tree to satisfy this is enormous. I suspect a deeper audit would be desirable before tackling that.
> > > For additional context, I was prototyping something with SharedString last night in a side project I'm working on, and am specifically using the fluid-framework package. In order to actually get my sequence delta event handler working, I had to dig into node_modules to figure out where I needed to get the type from. I don't think that's a reasonable user experience for our customers.
> > 
> > 
> > that makes sense, but it seems to me the re-export of shared string is just broken, and i'd suspect other re-export are similarly broken, and we'll continue to introduce new breaks, as the current strategy is not sustainable. the api-extractor thing seems like a good solution, and the surface area will be enormous. So i think we should either enable the api-extractor, or remove sequence from fluid-framework, as right now it doesn't really work
> 
> I'm actually wondering if Abram's suggestion to explicitly use type exports might reduce the API explosion I initially encountered when I tried the API extractor approach. Let me give that another shot and see if it is reasonable.
> 
> And I'm definitely happy to close this out if we want to take a different approach here - more than anything I just wanted to make sure this issue was on our radar 🙂

i guess i'm fine if you want to check this in, but i'd like to see a holistic plan for the fluid-framework package, rather than ad-hoc  changes, as i don't think we'll ever reach a good developer experience via ad-hoc changes

**Comment by tylerbutler** [comment]
@Josmithr @anthony-murphy #9050 is related.

**Comment by Josmithr** [comment]
> this is not unique to sequence. we use types across packages all the time in many of our packages, look at any definition package. if we think this pattern is bad, there is a whole lot more we need to do.

Wholeheartedly agree. @anthony-murphy @tylerbutler Do either of you know if there has been any existing discussion about this sort of broad package autonomy issue before? I'm happy to start a thread somewhere if not.

**Comment by anthony-murphy** [comment]
> > this is not unique to sequence. we use types across packages all the time in many of our packages, look at any definition package. if we think this pattern is bad, there is a whole lot more we need to do.
> 
> Wholeheartedly agree. @anthony-murphy @tylerbutler Do either of you know if there has been any existing discussion about this sort of broad package autonomy issue before? I'm happy to start a thread somewhere if not.

in general, i've found what we have works pretty well for per package consumption, and the tooling generally does a good job of helping me find things i need. i haven't been part of the fluid-framework package work, so i don't know what has been discussed there, and my guess is the re-export is what causes things to get lost.

**Comment by Josmithr** [comment]
I'm going to abandon this for now in favor of having a more comprehensive discussion about this. Additionally, work is being done in shared-string / merge-tree that will hopefully make the API-Extractor solution more tractable soon :)

**Pull request closed without merging**